Fix capture payments later

cesarcosta99 opened the pull request:
Fixes #2623 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This PR adds an implementation to pass on the capture method when creating a payment intent. On UPE, a payment intention is created when checkout loads and it was not including the capture method information that is saved in the settings. Because of that, all payments were being automatically captured.

The present changes fixes the manual capture issue on payments with new credit cards. It basically checks the settings and send that information to the Payments API.

If you have card + other payment methods enabled, you might notice that, when you have manual capture enabled, you will only see card as an option on checkout. The reason for that is simply because we're deliberately limiting the payment methods to card, as manual capture only works with it. There is an ongoing discussion on how to properly communicate that for the merchants and you can find more info on p1631843780047400-slack-CHG7MTCAF 

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
- Visit [the store](http://localhost:8082/) and log in as admin.
- In [WooCommerce Payments settings](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments), make sure `Issue an authorization on checkout, and capture later` option is unticked and saved.
- Go to the store > `Shop` and add a product to the cart.
- Go to checkout and pay using a new test card (e.g. 4242 4242 4242 4242).
- In the [Orders](http://localhost:8082/wp-admin/edit.php?post_type=shop_order), you should see the order you just placed with status `Processing`, as usual.
- Go to [WooCommerce Payments settings](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments) again, tick `Issue an authorization on checkout, and capture later` and save.
- Place a new order following the steps above.
- In the [Orders](http://localhost:8082/wp-admin/edit.php?post_type=shop_order), you should see the order you just placed with status `On hold` and a similar order note:
>A payment of €xx.xx was authorized using WooCommerce Payments (pi_...). Order status changed from Pending payment to On hold.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
-->

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

bborman22 reviewed (changes requested):
While this PR worked as described, one thing that I noticed was that we now have 2 different ways to check which payment methods are enabled at checkout and there was already a discrepancy. The original check was using `is_enabled_at_checkout` as an array filter and this new check is the `get_upe_enabled_at_checkout_payment_method_ids` function. 

With this PR there is a new divergence in the checks introduced, `is_enabled_at_checkout` has an additional currency check that is not captured in `get_upe_enabled_at_checkout_payment_method_ids`. To see the issue, set the currency to USD while having card and giropay enabled and manual capture set to no. This will cause an error trying to create the intent as giropay is not filtered out by currency.

I think it would be best if we could get all of these checks centralized in one place whether that is the array filter approach or it's own function, either would work just fine though if it is through the array filter, we should get the capture option once (in the constructor as a property?) so we're not getting it for each array element.

cesarcosta99 commented:
>I think it would be best if we could get all of these checks centralized in one place whether that is the array filter approach or it's own function, either would work just fine though if it is through the array filter, we should get the capture option once (in the constructor as a property?) so we're not getting it for each array element.

Good catch, @bborman22, thank you! I've kept `get_upe_enabled_at_checkout_payment_method_ids`, added the currency check to it, and replaced the remaining reference to `is_enabled_at_checkout` on 4a9a0be.
